### ISSUE TITLE
🐛 Running `npx rome init` goes in `heap out of memory`

### ISSUE THREAD

**Issue by alejandronanez** [rome/tools#1559]
I'm just trying out Rome for the first time, and if I run `npx rome init`, I get the message that I pasted bellow. I haven't installed Rome globally, it's not listed in my `package.json` either.

Also, I tried creating a new project:

```
npm init -y
npx rome init
```
And got the same error.

Thanks!!!

## Environment

```
$ rome rage --summary

Environment Variables
  Object {
    COLORFGBG: "15;0"
    COLORTERM: "truecolor"
    TERM: "xterm-256color"
    TERM_PROGRAM: "iTerm.app"
  }

User Config
  unset

Rome Version
  10.0.4-beta

Node Version
  14.17.0

Platform
  darwin x64 20.4.0

Terminal Features
  Object {
    background: "dark"
    colorDepth: 24
    columns: 122
    cursor: true
    hyperlinks: true
    isTTY: true
    unicode: true
  }

Client Flags
  Object {
    clientName: "cli"
    cwd: "/Users/alejandro/code/rome-101"
    realCwd: "/Users/alejandro/code/rome-101"
    silent: false
  }

Server Status
  Object {
    projects: Array [
      Object {id: 0}
      Object {id: 1}
    ]
    server: Object {
      heapTotal: 28_913_664
      pid: 9_086
      uptime: 1.567997975
    }
    workers: Array [
      Object {
        astCacheSize: 0
        heapTotal: 28_913_664
        ownedBytes: 0
        ownedFileCount: 0
        pid: 9_086
        uptime: 1.567912516
      }
    ]
  }
```

## Steps To Reproduce

1. On a machine that doesn't have `rome` installed globally, create a new directory and type `npx rome init`
2. It breaks with this message

```
<--- Last few GCs --->

[8807:0x104c40000]    62812 ms: Mark-sweep 4022.5 (4117.3) -> 4012.7 (4122.3) MB, 2463.6 / 0.6 ms  (average mu = 0.214, current mu = 0.055) allocation failure scavenge might not succeed
[8807:0x104c40000]    65391 ms: Mark-sweep 4028.7 (4122.3) -> 4017.1 (4127.5) MB, 2564.9 / 0.6 ms  (average mu = 0.118, current mu = 0.006) allocation failure scavenge might not succeed


<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x10130d6e5 node::Abort() (.cold.1) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
 2: 0x1000b1c49 node::Abort() [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
 3: 0x1000b1daf node::OnFatalError(char const*, char const*) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
 4: 0x1001f60f7 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
 5: 0x1001f6093 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
 6: 0x1003a54f5 v8::internal::Heap::FatalProcessOutOfMemory(char const*) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
 7: 0x1003a6fba v8::internal::Heap::RecomputeLimits(v8::internal::GarbageCollector) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
 8: 0x1003a2689 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
 9: 0x10039ff21 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
10: 0x1003ae7da v8::internal::Heap::AllocateRawWithLightRetrySlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
11: 0x1003ae861 v8::internal::Heap::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
12: 0x100376fdd v8::internal::FactoryBase<v8::internal::Factory>::NewRawOneByteString(int, v8::internal::AllocationType) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
13: 0x1004b285c v8::internal::JsonParser<unsigned char>::MakeString(v8::internal::JsonString const&, v8::internal::Handle<v8::internal::String>) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
14: 0x1004b112c v8::internal::JsonParser<unsigned char>::ParseJsonValue() [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
15: 0x1004b0622 v8::internal::JsonParser<unsigned char>::ParseJson() [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
16: 0x10029faf3 v8::internal::Builtin_Impl_JsonParse(v8::internal::BuiltinArguments, v8::internal::Isolate*) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
17: 0x100a894b9 Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_BuiltinExit [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
18: 0xe780b5108c9
[1]    8807 abort      npx rome init
```

## Expected Results

1. It should create a new rome project.

**Comment by alejandronanez** [comment]
Also, the same happens if I install `rome` doing `npm install rome` and then I create an `npm script` that runs `rome init`

```json
{
  "name": "rome-101",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1",
    "rome": "rome init"
  },
  "keywords": [],
  "author": "Alejandro Nanez",
  "license": "ISC",
  "dependencies": {
    "rome": "^10.0.4-beta"
  }
}
```

**Comment by ematipico** [comment]
I just tried it and it worked without issues. I used the same node version that you have.

Not sure how to replicate...

**Comment by alejandronanez** [comment]
@ematipico, I think I could try with a version different than LTS later today and report back.
Is there anything else you think I can try?

**Comment by JustBeYou** [comment]
@alejandronanez I found `rome` to be pretty resource intensive sometimes. I've got the same error a few times when I was really *abusing* my system. Could you check your RAM usage while running the command?

**Comment by alejandronanez** [comment]
I'm back with some updates:
@ematipico I tried with Node 16.3.0 and got the same error

```
Environment Variables
  Object {
    COLORFGBG: "15;0"
    COLORTERM: "truecolor"
    TERM: "xterm-256color"
    TERM_PROGRAM: "iTerm.app"
  }

User Config
  unset

Rome Version
  10.0.4-beta

Node Version
  14.17.0

Platform
  darwin x64 20.4.0

Terminal Features
  Object {
    background: "dark"
    colorDepth: 24
    columns: 78
    cursor: true
    hyperlinks: true
    isTTY: true
    unicode: true
  }

Client Flags
  Object {
    clientName: "cli"
    cwd: "/Users/alejandro/code/rome-101"
    realCwd: "/Users/alejandro/code/rome-101"
    silent: false
  }

Server Status
  Object {
    projects: Array [
      Object {id: 0}
      Object {id: 1}
    ]
    server: Object {
      heapTotal: 28_127_232
      pid: 13_094
      uptime: 0.235213577
    }
    workers: Array [
      Object {
        astCacheSize: 0
        heapTotal: 28_127_232
        ownedBytes: 0
        ownedFileCount: 0
        pid: 13_094
        uptime: 0.23515528
      }
    ]
 }
```

@JustBeYou Thanks for the hint! Actually `rome` has been using about 300%+ of my resources. I tried quiting every other application but `iTerm` and still got the same result 💔

At this point, I think I won't be able to use `rome` at all on my machine :( Is there anything you can think of to solve my issue?

Thanks!

![](https://cdn.zappy.app/c688382e4adebb387a95ab32d0811514.png)

**Comment by ematipico** [comment]
@alejandronanez could you trying using the latest nightly?

**Comment by alejandronanez** [comment]
> @alejandronanez could you trying using the latest nightly?

Hey @ematipico, thanks for the super quick response. Sadly I don't know how to do that, I tried with `nvm` but it seems that it doesn't have the option to install canaries. Which makes me assume that in order to run a Canary version of Node, I'd have to build it from source? If that's the case, then, I wouldn't be able to do that (I have time constrains).

**Comment by ematipico** [comment]
Hi @alejandronanez , sorry I meant the latest nightly of Rome. The version you used is quite old and maybe a recent one fixes your issue.

**Comment by alejandronanez** [comment]
> Hi @alejandronanez , sorry I meant the latest nightly of Rome. The version you used is quite old and maybe a recent one fixes your issue.

Oh gotcha @ematipico, let me try that out.

**Comment by alejandronanez** [comment]
@ematipico, I tried running the latest nightly, and I got pretty much the same error.

```
❯ npx rome@v10.0.7-nightly.2021.5.9 init
 Adding project ..                                                                                                                                              elapsed 44.29s
<--- Last few GCs --->

[45191:0x104c40000]    65746 ms: Mark-sweep (reduce) 4074.1 (4101.3) -> 4071.4 (4101.5) MB, 2281.2 / 1.6 ms  (average mu = 0.101, current mu = 0.039) allocation failure scavenge might not succeed
[45191:0x104c40000]    65754 ms: Scavenge (reduce) 4072.4 (4103.5) -> 4072.2 (4103.5) MB, 4.2 / 0.0 ms  (average mu = 0.101, current mu = 0.039) allocation failure


<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x10130d6e5 node::Abort() (.cold.1) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
 2: 0x1000b1c49 node::Abort() [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
 3: 0x1000b1daf node::OnFatalError(char const*, char const*) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
 4: 0x1001f60f7 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
 5: 0x1001f6093 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
 6: 0x1003a54f5 v8::internal::Heap::FatalProcessOutOfMemory(char const*) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
 7: 0x1003a6fba v8::internal::Heap::RecomputeLimits(v8::internal::GarbageCollector) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
 8: 0x1003a2689 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
 9: 0x10039ff21 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
10: 0x1003ae7da v8::internal::Heap::AllocateRawWithLightRetrySlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
11: 0x1003ae861 v8::internal::Heap::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
12: 0x10037bf12 v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationType, v8::internal::AllocationOrigin) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
13: 0x1006fc158 v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
14: 0x100a893d9 Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_NoBuiltinExit [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
[1]    45191 abort      npx rome@v10.0.7-nightly.2021.5.9 init
```

But, if I try with `v10.0.7-beta.8` I got this 🎉 (it's not completely working, but it's not the error that I've got so far)

```
❯ npx rome@v10.0.7-beta.8 init

 argv:1:25 args/invalid ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ✖ Missing arguments

    rome init --check-project
                             ^

  ℹ Expected exactly 1 arguments
```

**Comment by ematipico** [comment]
OK, thank you. So the issue is really around the `init` command. You could try to run the command passing the `--profile` flag. Doing so, Rome will generate a v8 profile and from there we might able to understand what's the issue.

**Comment by alejandronanez** [comment]
Thanks @ematipico!

Here's the output I got

```
~/code/rome-101 is 📦 v1.0.0 via ⬢ v14.17.0
❯ npx rome@v10.0.7-beta.8 init --profile
ℹ Starting CPU profile...

 argv:1:25 args/invalid ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ✖ Missing arguments

    rome init --check-project
                             ^

  ℹ Expected exactly 1 arguments

(node:46869) UnhandledPromiseRejectionWarning: DiagnosticsError: Resource released

 /Users/alejandro/.npm/_npx/46869/lib/node_modules/rome/bin/rome/index.js:124749:4 bridge/closed  INTERNAL

  ✖ Resource released

   1. events.end (/Users/alejandro/.npm/_npx/46869/lib/node_modules/rome/bin/rome/index.js:124749:4)
   2. finalize (/Users/alejandro/.npm/_npx/46869/lib/node_modules/rome/bin/rome/index.js:124367:17)
   3. /Users/alejandro/.npm/_npx/46869/lib/node_modules/rome/bin/rome/index.js:125762:11
   4. new Promise
   5. callRelease (/Users/alejandro/.npm/_npx/46869/lib/node_modules/rome/bin/rome/index.js:125761:22)
   6. resources._release (/Users/alejandro/.npm/_npx/46869/lib/node_modules/rome/bin/rome/index.js:125701:11)
   7. Function.all
   8. resources._release (/Users/alejandro/.npm/_npx/46869/lib/node_modules/rome/bin/rome/index.js:125690:4)
   9. client.end (/Users/alejandro/.npm/_npx/46869/lib/node_modules/rome/bin/rome/index.js:153623:3)
  10. cli (/Users/alejandro/.npm/_npx/46869/lib/node_modules/rome/bin/rome/index.js:166890:3)

  ⚠ This diagnostic was derived from an internal Rome error. Potential bug, please report if necessary.


  at createSingleDiagnosticsError (/Users/alejandro/.npm/_npx/46869/lib/node_modules/rome/bin/rome/index.js:119412:9)
  at createRuntimeDiagnosticsError (/Users/alejandro/.npm/_npx/46869/lib/node_modules/rome/bin/rome/index.js:119404:9)
  at events.end (/Users/alejandro/.npm/_npx/46869/lib/node_modules/rome/bin/rome/index.js:124749:4)
  at finalize (/Users/alejandro/.npm/_npx/46869/lib/node_modules/rome/bin/rome/index.js:124367:17)
  at <anonymous> (/Users/alejandro/.npm/_npx/46869/lib/node_modules/rome/bin/rome/index.js:125762:11)
  at new Promise
  at callRelease (/Users/alejandro/.npm/_npx/46869/lib/node_modules/rome/bin/rome/index.js:125761:22)
  at resources._release (/Users/alejandro/.npm/_npx/46869/lib/node_modules/rome/bin/rome/index.js:125701:11)
  at Function.all
  at resources._release (/Users/alejandro/.npm/_npx/46869/lib/node_modules/rome/bin/rome/index.js:125690:4)
  at client.end (/Users/alejandro/.npm/_npx/46869/lib/node_modules/rome/bin/rome/index.js:153623:3)
  at cli (/Users/alejandro/.npm/_npx/46869/lib/node_modules/rome/bin/rome/index.js:166890:3)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:46869) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 4)
(node:46869) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
/Users/alejandro/.npm/_npx/46869/lib/node_modules/rome/bin/rome/index.js:125589
							throw new Error(
							^

[Error: The resource ReporterProgress is not correctly managed as it has not been attached to a parent resource. Set the environment variable ROME_RESOURCE_STACKS=1 for a creation stack.
  at Timeout._onTimeout (/Users/alejandro/.npm/_npx/46869/lib/node_modules/rome/bin/rome/index.js:125589:13)
  at listOnTimeout (internal/timers.js:555:16)
  at process.processTimers (internal/timers.js:498:6)]
```

**Comment by ematipico** [comment]
@alejandronanez could you please do it using the latest nightly?

**Comment by alejandronanez** [comment]
@ematipico there you go

```
❯ npx rome@v10.0.7-nightly.2021.5.9 init --profile
ℹ Starting CPU profile...
 Adding project ..                                                                                                                                                           elapsed 45.30s
<--- Last few GCs --->

[47491:0x104d40000]    63641 ms: Mark-sweep (reduce) 4070.3 (4101.3) -> 4069.2 (4104.0) MB, 2381.1 / 1.9 ms  (average mu = 0.114, current mu = 0.005) allocation failure scavenge might not succeed
[47491:0x104d40000]    66030 ms: Mark-sweep (reduce) 4070.2 (4104.5) -> 4069.6 (4104.8) MB, 2385.1 / 1.5 ms  (average mu = 0.060, current mu = 0.002) allocation failure scavenge might not succeed


<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x10130d6e5 node::Abort() (.cold.1) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
 2: 0x1000b1c49 node::Abort() [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
 3: 0x1000b1daf node::OnFatalError(char const*, char const*) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
 4: 0x1001f60f7 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
 5: 0x1001f6093 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
 6: 0x1003a54f5 v8::internal::Heap::FatalProcessOutOfMemory(char const*) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
 7: 0x1003a6fba v8::internal::Heap::RecomputeLimits(v8::internal::GarbageCollector) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
 8: 0x1003a2689 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
 9: 0x10039ff21 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
10: 0x1003ae7da v8::internal::Heap::AllocateRawWithLightRetrySlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
11: 0x1003ae861 v8::internal::Heap::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
12: 0x10037bf12 v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationType, v8::internal::AllocationOrigin) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
13: 0x1006fc158 v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
14: 0x100a893d9 Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_NoBuiltinExit [/Users/alejandro/.nvm/versions/node/v14.17.0/bin/node]
[1]    47491 abort      npx rome@v10.0.7-nightly.2021.5.9 init --profile
```

**Comment by jamiebuilds** [comment]
Hi! Thank you for opening this issue

We recently made the decision to [migrate the Rome codebase to Rust](https://rome.tools/blog/2021/09/21/rome-will-be-rewritten-in-rust) and making some significant changes to its architecture. We've also decided not to release the previous version written in TypeScript.

We're going through and closing pull requests and issues that were specific to the TypeScript codebase, including this one. Thank you for taking the time to open it.

Thanks again!
